### PR TITLE
[DEPRECATE] - RFC 421 - deprecate app ctrl router props

### PR DESCRIPTION
**Pull request by mistahenry**
Implements [RFC #421](https://github.com/emberjs/rfcs/pull/421/files).
Part of https://github.com/emberjs/ember.js/issues/17569.

Please let me know if `notifyPropertyChange` was an okay way to solve this.

### CODE REVIEW

**Comment by locks** [comment]
Can you add the following URL to the deprecation, `https://emberjs.com/deprecations/v3.x#toc_application-controller-router-properties`?
It should become valid when https://github.com/ember-learn/deprecation-app/pull/276 gets merged.

**Comment by mistahenry** [comment]
@locks Updated the deprecation with the url/ matching deprecation id

File: packages/@ember/-internals/routing/lib/system/router.ts [hunk 1]
```diff
@@ -1637,16 +1637,44 @@ function updatePaths(router: EmberRouter) {
...
+    });
   }
-
-  set(appController, 'currentPath', path);
+  notifyPropertyChange(appController, 'currentPath');
 
   if (!('currentRouteName' in appController)) {
-    defineProperty(appController, 'currentRouteName');
+    Object.defineProperty(appController, 'currentRouteName', {
+      enumerable: false,
```
Comment (chadhietala) [inline packages/@ember/-internals/routing/lib/system/router.ts]: This shouldn't be needed.

File: packages/@ember/-internals/routing/lib/system/router.ts [hunk 2]
```diff
@@ -1637,16 +1637,44 @@ function updatePaths(router: EmberRouter) {
   }
 
   if (!('currentPath' in appController)) {
-    defineProperty(appController, 'currentPath');
+    Object.defineProperty(appController, 'currentPath', {
+      enumerable: false,
```
Comment (chadhietala) [inline packages/@ember/-internals/routing/lib/system/router.ts]: This shouldn't be needed.

File: packages/@ember/-internals/routing/lib/system/router.ts [hunk 3]
```diff
@@ -1637,16 +1637,44 @@ function updatePaths(router: EmberRouter) {
   }
 
   if (!('currentPath' in appController)) {
```
Comment (chadhietala) [inline packages/@ember/-internals/routing/lib/system/router.ts]: We need to wrap all the code that is set for deprecation in a deprecation flag. This allows you to build a version of Ember without any of the deprecated code. If you need an example of what you need to do I would look at. https://github.com/emberjs/ember.js/pull/17193

**Review by chadhietala — CHANGES_REQUESTED** [review]
This looks great! Thanks for doing this, just a couple small things we need to address before merging.

File: packages/@ember/-internals/routing/lib/system/router.ts (hunk 3 above)
Comment (mistahenry) [reply packages/@ember/-internals/routing/lib/system/router.ts]: I chose `export const APP_CTRL_ROUTER_PROPS = !!'3.9.0';` but that was only based on [the deprecation link](https://github.com/ember-learn/deprecation-app/pull/276/files) and my assumption that `3.9.0` refers to the version in which this is deprecated and not when the deprecation is removed.

Comment (chadhietala) [reply packages/@ember/-internals/routing/lib/system/router.ts]: It's actually when it needs to be removed so it would need to be `!!'4.0.0'`.

Comment (mistahenry) [reply packages/@ember/-internals/routing/lib/system/router.ts]: alright, I've updated it

**Comment by chadhietala** [comment]
Thanks so much @mistahenry!

**Pull request merged**